src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -3908,13 +3910,14 @@ private static string MatchCharacterClass(bool hasTextInfo, RegexOptions options
             // it may be cheaper and smaller to compare against each than it is to use a lookup table.  We can also special-case
             // the very common case with case insensitivity of two characters next to each other being the upper and lowercase
             // ASCII variants of each other, in which case we can use bit manipulation to avoid a comparison.
-            if (!invariant && !RegexCharClass.IsNegated(charClass))
+            if (!invariant && (!caseInsensitive || !RegexCharClass.IsNegated(charClass)))
```

Comment:
Would it be worth it to call RegexCharClass.GetSetChars before the ifcheck in order to be able to add to the if clause an additional `... && setChars.Count <= 3)` ?